File: algorithms/geometry/closest-pair-of-points/typescript/closestPair.ts
export function closestPair(arr: number[]): number {
    const n = arr.length / 2;
    const points: [number, number][] = [];
    for (let i = 0; i < n; i++) {
        points.push([arr[2 * i], arr[2 * i + 1]]);
    }
    points.sort((a, b) => a[0] !== b[0] ? a[0] - b[0] : a[1] - b[1]);

    function distSq(a: [number, number], b: [number, number]): number {
        return (a[0] - b[0]) * (a[0] - b[0]) + (a[1] - b[1]) * (a[1] - b[1]);
    }

    function solve(l: number, r: number): number {
        if (r - l < 3) {
            let min = Infinity;
            for (let i = l; i <= r; i++) {
                for (let j = i + 1; j <= r; j++) {
                    min = Math.min(min, distSq(points[i], points[j]));
                }
            }
            return min;
        }

        const mid = Math.floor((l + r) / 2);
        const midX = points[mid][0];

        const dl = solve(l, mid);
        const dr = solve(mid + 1, r);
        let d = Math.min(dl, dr);

        const strip: [number, number][] = [];
        for (let i = l; i <= r; i++) {
            if ((points[i][0] - midX) * (points[i][0] - midX) < d) {
                strip.push(points[i]);
            }
        }
        strip.sort((a, b) => a[1] - b[1]);

        for (let i = 0; i < strip.length; i++) {
            for (let j = i + 1; j < strip.length &&
                    (strip[j][1] - strip[i][1]) * (strip[j][1] - strip[i][1]) < d; j++) {
                d = Math.min(d, distSq(strip[i], strip[j]));
            }
        }

        return d;
    }

    return solve(0, n - 1);
}

File: algorithms/geometry/closest-pair-of-points/java/ClosestPair.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;

public class ClosestPair {

    public static int closestPair(int[] arr) {
        int n = arr.length / 2;
        int[][] points = new int[n][2];
        for (int i = 0; i < n; i++) {
            points[i][0] = arr[2 * i];
            points[i][1] = arr[2 * i + 1];
        }
        Arrays.sort(points, (a, b) -> a[0] != b[0] ? a[0] - b[0] : a[1] - b[1]);
        return solve(points, 0, n - 1);
    }

    private static int distSq(int[] p1, int[] p2) {
        return (p1[0] - p2[0]) * (p1[0] - p2[0]) + (p1[1] - p2[1]) * (p1[1] - p2[1]);
    }

    private static int solve(int[][] points, int l, int r) {
        if (r - l < 3) {
            int min = Integer.MAX_VALUE;
            for (int i = l; i <= r; i++) {
                for (int j = i + 1; j <= r; j++) {
                    min = Math.min(min, distSq(points[i], points[j]));
                }
            }
            return min;
        }

        int mid = (l + r) / 2;
        int midX = points[mid][0];

        int dl = solve(points, l, mid);
        int dr = solve(points, mid + 1, r);
        int d = Math.min(dl, dr);

        List<int[]> strip = new ArrayList<>();
        for (int i = l; i <= r; i++) {
            if ((points[i][0] - midX) * (points[i][0] - midX) < d) {
                strip.add(points[i]);
            }
        }
        strip.sort((a, b) -> a[1] - b[1]);

        for (int i = 0; i < strip.size(); i++) {
            for (int j = i + 1; j < strip.size() &&
                    (strip.get(j)[1] - strip.get(i)[1]) * (strip.get(j)[1] - strip.get(i)[1]) < d; j++) {
                d = Math.min(d, distSq(strip.get(i), strip.get(j)));
            }
        }

        return d;
    }
}
